File: python_solutions/1578.minimum-deletion-cost-to-avoid-repeating-letters.py
from collections import Counter, defaultdict, OrderedDict, deque
from bisect import bisect_left, bisect_right
from functools import reduce, lru_cache
from typing import List
import itertools
import math
import heapq
import string
true = True
false = False
MIN, MAX, MOD = -0x3f3f3f3f, 0x3f3f3f3f, 1000000007


#
# @lc app=leetcode id=1578 lang=python3
#
# [1578] Minimum Deletion Cost to Avoid Repeating Letters
#
# https://leetcode.com/problems/minimum-deletion-cost-to-avoid-repeating-letters/description/
#
# algorithms
# Medium (56.92%)
# Total Accepted:    5.5K
# Total Submissions: 9.7K
# Testcase Example:  '"abaac"\n[1,2,3,4,5]'
#
# Given a string s and an array of integers cost where cost[i] is the cost of
# deleting the character i in s.
#
# Return the minimum cost of deletions such that there are no two identical
# letters next to each other.
#
# Notice that you will delete the chosen characters at the same time, in other
# words, after deleting a character, the costs of deleting other characters
# will not change.
#
#
# Example 1:
#
#
# Input: s = "abaac", cost = [1,2,3,4,5]
# Output: 3
# Explanation: Delete the letter "a" with cost 3 to get "abac" (String without
# two identical letters next to each other).
#
#
# Example 2:
#
#
# Input: s = "abc", cost = [1,2,3]
# Output: 0
# Explanation: You don't need to delete any character because there are no
# identical letters next to each other.
#
#
# Example 3:
#
#
# Input: s = "aabaa", cost = [1,2,3,4,1]
# Output: 2
# Explanation: Delete the first and the last character, getting the string
# ("aba").
#
#
#
# Constraints:
#
#
# s.length == cost.length
# 1 <= s.length, cost.length <= 10^5
# 1 <= cost[i] <= 10^4
# s contains only lowercase English letters.
#
#
#
class Solution:
    def minCost(self, s: str, cost: List[int]) -> int:
        i, res, n = 0, 0, len(cost)
        while i < n - 1:
            if s[i] == s[i + 1]:
                j, total, n_max = i, 0, 0
                while j < n and s[i] == s[j]:
                    total += cost[j]
                    n_max = max(n_max, cost[j])
                    j += 1
                res += total - n_max
                i = j
            else:
                i += 1
        return res


sol = Solution()

s, cost = "abaac", [1, 2, 3, 4, 5]
s, cost = "abc", [1, 2, 3]
s, cost = "aabaa", [1, 2, 3, 4, 1]
print(sol.minCost(s, cost))

File: python_solutions/1578.minimum-deletion-cost-to-avoid-repeating-letters.cpp
// C libraries
#include <cassert>
#include <cctype>
#include <climits>
#include <cmath>
#include <cstdio>
#include <cstdlib>
#include <cstring>
#include <ctime>

// Containers
#include <deque>
#include <list>
#include <map>
#include <queue>
#include <set>
#include <stack>
#include <unordered_map>
#include <unordered_set>
#include <vector>

// Input/Output
#include <fstream>
#include <iomanip>
#include <ios>
#include <iostream>
#include <istream>
#include <ostream>
#include <sstream>

// Other
#include <algorithm>
#include <bitset>
#include <chrono>
#include <iterator>
#include <limits>
#include <random>
#include <stdexcept>
#include <string>
#include <tuple>
#include <type_traits>
#include <utility>

// ==================================================

using namespace std;

// constants
const double EPS = 1e-9;
const int MOD = 1000000007;

// type alias
using pii = pair<int, int>;
using ll = long long;
using ull = unsigned long long;
using vi = vector<int>;
using vb = vector<bool>;
using vc = vector<char>;
using vs = vector<string>;
using vvb = vector<vector<bool>>;
using vvc = vector<vector<char>>;
using vvi = vector<vector<int>>;
using vvs = vector<vector<string>>;
using mii = map<int, int>;
using mci = map<char, int>;
using si = set<int>;
using spii = set<pair<int, int>>;
using umii = unordered_map<int, int>;
using umci = unordered_map<char, int>;
using umsi = unordered_map<string, int>;
using usi = unordered_set<int>;
using usc = unordered_set<char>;
using uss = unordered_set<string>;
using vpii = vector<pair<int, int>>;

typedef struct TreeNode TreeNode;
using ptn = TreeNode *;
using tn = TreeNode;

// ==================================================

// fast IO
static auto __speedup__ = []() {
  ios::sync_with_stdio(false);
  cin.tie(nullptr);
  return 0;
}();

// ==================================================

// some macro for less typing
#define forloop(i, n) for (int i = 0; i < n; i++)             //[0, n)
#define forloopr(i, n) for (int i = n - 1; i >= 0; --i)       // reverse [0, n)
#define forloopup(i, a, b) for (int i = a; i < b; ++i)        // [a, b)
#define forloopdown(i, a, b) for (int i = b - 1; i >= a; --i) // reverse [a, b)
#define forunfold(i, arr) for (auto &i : arr)

#define INF 0x3f3f3f3f
#define MAX(a, b) a = max(a, b)
#define MIN(a, b) a = min(a, b)

#define fi first
#define se second
#define mp make_pair
#define pb push_back
#define pf push_front
#define ef emplace_front
#define eb emplace_back
#define qall(v) v.begin(), v.end()
#define qsize(v) (int)v.size()
#define qsort(v) sort(v.begin(), v.end())
#define rqsort(v) sort(v.rbegin(), v.rend())
#define qreverse(v) reverse(v.begin(), v.end())

// int to string
string itos(int n) { return to_string(n); }
// char to string
string ctos(char c) { return string(1, c); };

inline string upper(string s) {
  string t(s);
  transform(t.begin(), t.end(), t.begin(), ::toupper);
  return t;
}
inline string lower(string s) {
  string t(s);
  transform(t.begin(), t.end(), t.begin(), ::tolower);
  return t;
}

int dirs[5] = {-1, 0, 1, 0, -1};

// ==================================================
/*
 * @lc app=leetcode id=1578 lang=cpp
 *
 * [1578] Minimum Deletion Cost to Avoid Repeating Letters
 *
 * https://leetcode.com/problems/minimum-deletion-cost-to-avoid-repeating-letters/description/
 *
 * algorithms
 * Medium (56.92%)
 * Total Accepted:    5.5K
 * Total Submissions: 9.7K
 * Testcase Example:  '"abaac"\n[1,2,3,4,5]'
 *
 * Given a string s and an array of integers cost where cost[i] is the cost of
 * deleting the character i in s.
 *
 * Return the minimum cost of deletions such that there are no two identical
 * letters next to each other.
 *
 * Notice that you will delete the chosen characters at the same time, in other
 * words, after deleting a character, the costs of deleting other characters
 * will not change.
 *
 *
 * Example 1:
 *
 *
 * Input: s = "abaac", cost = [1,2,3,4,5]
 * Output: 3
 * Explanation: Delete the letter "a" with cost 3 to get "abac" (String without
 * two identical letters next to each other).
 *
 *
 * Example 2:
 *
 *
 * Input: s = "abc", cost = [1,2,3]
 * Output: 0
 * Explanation: You don't need to delete any character because there are no
 * identical letters next to each other.
 *
 *
 * Example 3:
 *
 *
 * Input: s = "aabaa", cost = [1,2,3,4,1]
 * Output: 2
 * Explanation: Delete the first and the last character, getting the string
 * ("aba").
 *
 *
 *
 * Constraints:
 *
 *
 * s.length == cost.length
 * 1 <= s.length, cost.length <= 10^5
 * 1 <= cost[i] <= 10^4
 * s contains only lowercase English letters.
 *
 *
 */
class Solution {
public:
  int minCost(string s, vector<int> &cost) {
    int i = 0, res = 0, n = s.size();
    while (i < n - 1) {
      if (s[i] == s[i + 1]) {
        int total = 0, j = i, n_max = 0;
        while (j < n && s[i] == s[j])
          total += cost[j], n_max = max(n_max, cost[j]), j++;
        res += total - n_max;
        i = j;
      } else {
        i++;
      }
    }
    return res;
  }
};

auto speed_up = []() {
  ios_base::sync_with_stdio(false);
  return 0;
}();
